README.md
# jsignage
Jewish Signage for Beit Knesset, in Hebrew, based on java script


text/news.txt
מזל טוב לאיתי דיין להגעתו לגיל מצוות. מזל טוב להורים ולכל המשפחה

בית מדרש עם הרב ברקוביץ יום שישי לאחר התפילה

date/GLOBAL.js
function c(){console.log(c.arguments)}
if (!(window.console)){
var console = {
	log:function(){ 
		for (var console_arg=0;console_arg <arguments.length;console_arg++){
			$("log").innerHTML = $("log").innerHTML + arguments[console_arg]
		}}}}
console.print = function(){
	$("log").innerHTML +=  "<br />"
	for (var console_arg=0;console_arg <arguments.length;console_arg++){
			$("log").innerHTML = $("log").innerHTML +","+ arguments[console_arg]
	}	
}
function $(id){return document.getElementById(id)}
function cursorPosition(ev){ev = ev || window.event;	
	if(ev.pageX || ev.pageY){return {x:ev.pageX, y:ev.pageY};}
	return {x:ev.clientX + document.body.scrollLeft - document.body.clientLeft,
		y:ev.clientY + document.body.scrollTop  - document.body.clientTop};
}


var Cookie = {
  build: function() {
	 arr_b = new Array()
	for (var i=0;i<arguments.length;i++){
		arr_b.push(arguments[i])
	}
    return arr_b.join("; ");
  },
  daysFromNow: function(days) {
    var d = new Date();
    d.setTime(d.getTime() +(days * 24 * 60 * 60 * 1000));
    return d.toGMTString();
  },
  set: function(name,value,day){
    var expiry = day ? 'expires=' + Cookie.daysFromNow(day) : null;
    document.cookie = Cookie.build(name + "=" + value, expiry, "path=/");
  }
  ,
  get: function(name){
    var valueMatch = new RegExp(name + "=([^;]+)").exec(document.cookie);
    return valueMatch ? valueMatch[1] : "";
  },
  del: function(name){
    Cookie.set(name,'',-1);
  }
};


function Move(e,elm){//Move by tlissak v 0.7 
	var elm = elm
	elm.style.position = "absolute"
	var process 	= true
	var cursorPosition = function (ev){
		ev = ev || window.event;
		if(ev.pageX || ev.pageY){			return {x:ev.pageX, y:ev.pageY};		}
		return {x:ev.clientX + document.body.scrollLeft - document.body.clientLeft,
				y:ev.clientY + document.body.scrollTop  - document.body.clientTop}
	}
	var findObjPos = function (obj) {
		if (obj){var curleft = 0 ;var curtop = 0;
			if (obj.offsetParent) {
				curleft = obj.offsetLeft ; curtop = obj.offsetTop
				while (obj = obj.offsetParent) {curleft += obj.offsetLeft ; curtop += obj.offsetTop	}}
			return {x:curleft,y:curtop};
		}
	}	
	var old_y_pos  = parseInt(elm.style.top) ? parseInt(elm.style.top) : findObjPos(elm).y
	var old_x_pos	= parseInt(elm.style.left) ? parseInt(elm.style.left) : findObjPos(elm).x
	var old_csr_x  = cursorPosition(e).x - old_x_pos
	var old_csr_y	= cursorPosition(e).y - old_y_pos
	
	document.onmousemove= mousemove
	document.body.onselectstart=function(){return false}//ie
	document.body.onmousedown=function(){return false}//mozilla}
	function mouseup(){
		process=false
		document.onmouseup = null
		document.onmousemove = null
		document.body.onselectstart=function(){return true}//ie
		document.body.onmousedown=function(){return true}//mozilla
	}
	function mousemove(e){
		elm.onmouseup = mouseup
		document.onmouseup = mouseup
		if(process){
			elm.style.left = cursorPosition(e).x - old_csr_x +"px"
			elm.style.top  = cursorPosition(e).y - old_csr_y +"px"
			document.onmouseup = this.mouseup
		}
	}	
}

date/ZMANIM.js
/* ksun.js - Kaluach suntimes Javascript routines
 *   Version 1.00 (initial release)
 *   Version 1.01 (fixed bug with time adjust - AM/PM and 24 hour clock)
 *   Version 1.02 (fixed bug with time adjust [again] - AM/PM and 24 hour clock)
 *   Version 2.00 (new suntimes routine, original routine was buggy)
 *   Version 2.01 (handle invalid sunrise/set, different knissat shabbat times)
 * Copyright (C) 5760-5763 (2000 - 2003 CE), by Abu Mami and Yisrael Hersch.
 *   All Rights Reserved.
 *   All copyright notices in this script must be left intact.
 * Based on:
 *   - PHP code that was translated by mattf@mail.com from the original perl
 *     module Astro-SunTime-0.01
 *       - original version of ksun.js was based on the program SUN.C by Michael
 *     Schwartz, which was based on an algorithm contained in:
 *         Almanac for Computers, 1990
 *         published by Nautical Almanac Office
 *         United States Naval Observatory
 *         Washington, DC 20392
 * Permission will be granted to use this script on your web page
 *   if you wish. All that's required is that you please ask.
 *   (Of course if you want to send a few dollars, that's OK too :-)
 * website: http://www.kaluach.net
 * email: abumami@kaluach.org
 * Edited by tlissak@gmail.com
 */
//OP = { lond:2,lonm: 20, ns:0/*N*/,latd: 48,latm: 50,ew: 1/*E*/, tz:1,stdi:14,place:"Paris"} //???? not verified
//OP = { lond:37,lonm: 37, ns:0/*N*/,latd: 55,latm: 45,ew: 1/*E*/, tz:3,stdi:16,place:"Moscow"}
/* 
.getZmanim() : available times :
city,hanetz,shkia,shaa_zmanit,alot,misheyakir,tzeit,shema,tefillah,chatzot,minchag,minchak,plag,motzeiShabbat,knissatShabbat
*/
//TEST : moscow : OP = { lond:37,lonm: 37, ns:0/*N*/,latd: 55,latm: 45,ew: 1/*E*/, tz:3,stdi:16,place:"Moscow"}
/*
[3, 8, "2009", 90, 50, 37, 37, 1, 55, 45, 0, 3]
[3, 8, "2009", 106, 6, 37, 37, 1, 55, 45, 0, 3]
[3, 8, "2009", 101, 0, 37, 37, 1, 55, 45, 0, 3]
[3, 8, "2009", 96, 0, 37, 37, 1, 55, 45, 0, 3]
results
128,258,440,837,957,1235,1315,1713,1852,2031,2117 */
//TEST : jerusalem : OP = { lond:35,lonm: 14, ns:0/*N*/,latd: 31,latm: 46,ew: 1/*E*/, tz:2,stdi:15,place:"Jerusalem"}
/*[3, 8, "2009", 90, 50, 35, 14, 1, 31, 46, 0, 2]
[3, 8, "2009", 106, 6, 35, 14, 1, 31, 46, 0, 2]
[3, 8, "2009", 101, 0, 35, 14, 1, 31, 46, 0, 2]
[3, 8, "2009", 96, 0, 35, 14, 1, 31, 46, 0, 2]
results :
Alot Hashachar (dawn)  	 336 , Misheyakir 	404 , Hanetz Hachama (sunrise) 	456,Sof  zman  shema (gr"a) 	821
Sof  zman  tefillah (gr"a) 	929,Chatzot  hayom (noon) 	1145,Mincha  gedolah 	 1219,Mincha  ketana  	1543
Plag  hamincha 	1708 , Shkiat  hachama (sunset) 	1833,Tzeit  hakochavim (nightfall) 	1900,Knissat  Shabbat ,Motzei  Shabbat
*/
JDate.prototype.getZmanim = function(_O,_timedmc){
		if( typeof(_O) != "object" ){ 
			_O = { lond:35,lonm: 14, ns:0/*N*/,latd: 31,latm: 46,ew: 1/*E*/, tz:2,stdi:15,place:"Jerusalem"}
			console.log("getZmanim(_O) _O is!object set Jerusalem ")
		}
		
		var place = _O.place.toLowerCase()
		
		g 	=  (this.Class == GDate) ? 	this : new GDate(this)//alwayes convert this to gregorien date
		
		
		var _D = {d:g.getDay(),m:g.getMonth(),y:g.getYear()}
		function timeadj(t) {
			var hour;
			var min;
			var time = t;
			var hour = Math.floor(time);
			var min  = Math.floor((time - hour) * 60.0 + 0.5);
			if(min >= 60) { hour += 1;  min  -= 60;  }
			if(hour < 0)
					hour += 24;
			var str = hour + ':' + ((min < 10) ? '0' : '') + min ;
			return str;
		}	
		
		time = suntime(_D, 90, 50, _O,_timedmc);
		sunrise = time.sr; // 2 sr
		sunset  = time.ss; // 3 ss
		hanetz = timeadj(sunrise);
		shkia = timeadj(sunset);
		shaa_zmanit = (sunset - sunrise) / 12;
		time = suntime(_D, 106, 6, _O,_timedmc);
		alot = timeadj(time.sr);
		time = suntime(_D, 101, 0,_O,_timedmc);
		misheyakir = timeadj(time.sr);
		time = suntime(_D, 96, 0, _O,_timedmc);
		tzeit = timeadj(time.ss);
		shema    = timeadj(sunrise + shaa_zmanit * 3);
		tefillah = timeadj(sunrise + shaa_zmanit * 4);
		chatzot  = timeadj(sunrise + shaa_zmanit * 6);
		minchag  = timeadj(sunrise + shaa_zmanit * 6.5);
		minchak  = timeadj(sunrise + shaa_zmanit * 9.5);
		plag     = timeadj(sunrise + shaa_zmanit * 10.75);
		time = suntime(_D, 98, 30, _O,_timedmc);// motzei shabbat (3 small stars)
		motzeiShabbat = timeadj(time.ss);
		var day_before // get knisat shabat	for day before
		if (place == "jerusalem")
			xmns = (40.0/ 60.0)
		else if(place == "haifa" || place == "be'er Sheva")
			xmns = (30.0/ 60.0)
		else if(place == "karnei Shomron" || place == "tel Aviv")
			xmns = (22.0/ 60.0)
		else{ 
			xmns = (18.0/ 60.0)
		}
		knissatShabbat = timeadj(sunset  - xmns);	
		return {
			city:_O
			,sunrise:sunrise,sunset:sunset
			,hanetz:hanetz,	shkia:shkia ,shaa_zmanit:shaa_zmanit,alot:alot,misheyakir:misheyakir
			,tzeit:tzeit,shema:shema,tefillah:tefillah,chatzot:chatzot,minchag:minchag
			,minchak:minchak,plag:plag
			,motzeiShabbat:motzeiShabbat
			,knissatShabbat:knissatShabbat
		}
	function suntime(  _D /*{d:01,m:01,y:2001}gregorien date*/, sundeg , sunmin, _OPOS ,P_timedmc){
			function doy(d, m, y) {return monCount[m] + d + (m > 2 && leap(y));}
			function leap(y) { return ((y % 400 == 0) || (y % 100 != 0 && y % 4 == 0));}
			var monCount = new makeArray(1, 32, 60, 91, 121, 152, 182, 213, 244, 274, 305, 335, 366);
			function makeArray() {
				this[0] = makeArray.arguments.length;
					for (i = 0; i < makeArray.arguments.length; i = i + 1)
						this[i+1] = makeArray.arguments[i];
			}
			var invalid = 0;        // start out as OK
			var ss;
			var sr;
			dy = _D.d
			mn = _D.m
			yr = String(_D.y)
			londeg 	= _OPOS.lond
			lonmin 	= _OPOS.lonm
			ew 		= _OPOS.ew
			latdeg 	= _OPOS.latd
			latmin 	= _OPOS.latm
			ns 		= _OPOS.ns
			

			_timezone = _OPOS.tz	+ P_timedmc
			
			longitude = (londeg + lonmin/60.0) * ((ew == 0) ? -1 : 1);
			latitude  = (latdeg + latmin/60.0) * ((ns == 0) ? 1 : -1);
			var yday = doy(dy, mn, yr);
			var A = 1.5708; 
			var B = 3.14159; 
			var C = 4.71239; 
			var D = 6.28319;      
			var E = 0.0174533 * latitude; 
			var F = 0.0174533 * longitude; 
			var G = 0.261799 * _timezone; 
			var R = Math.cos(0.01745 * (sundeg + sunmin/60.0));
			var J;
			/* two times through the loop   i=0 is for sunrise    i=1 is for sunset */
			for (var i = 0; i < 2; i++) { 
					if(!i)
							J =  A; // sunrise 
					else
							J = C;  // sunset
					var K = yday + ((J - F) / D); 
					var L = (K * .017202) - .0574039;              // Solar Mean Anomoly 
					var M = L + .0334405 * Math.sin(L);            // Solar True Longitude 
					M += 4.93289 + (3.49066E-04) * Math.sin(2 * L); 
					// Quadrant Determination 
					if (D == 0) {
							alert("Trying to normalize with zero offset..."); exit;
					}
					while(M < 0)
							M = (M + D);
					while(M >= D)
							M = (M - D);
					if ((M / A) - Math.floor(M / A) == 0)
							M += 4.84814E-06;
					var P = Math.sin(M) / Math.cos(M);                   /// Solar Right Ascension 
					P = Math.atan2(.91746 * P, 1); 
					// Quadrant Adjustment 
					if (M > C)
							P += D;
					else {
							if (M > A)
									P += B;
					} 
					var Q = .39782 * Math.sin(M);      // Solar Declination 
					Q = Q / Math.sqrt(-Q * Q + 1);     /// This is how the original author wrote it! 
					Q = Math.atan2(Q, 1); 
					var S = R - (Math.sin(Q) * Math.sin(E)); 
					S = S / (Math.cos(Q) * Math.cos(E)); 
					if(Math.abs(S) > 1)
							invalid = 1;    // uh oh! no sunrise/sunset
					S = S / Math.sqrt(-S * S + 1); 
					S = A - Math.atan2(S, 1); 
					if(!i)
							S = D - S;      // sunrise
					var T = S + P - 0.0172028 * K - 1.73364;  // Local apparent time 
					var U = T - F;                            // Universal timer 
					var V = U + G;                            // Wall clock time 
					// Quadrant Determination 
					if(D == 0) {
							alert("Trying to normalize with zero offset...");   exit;
					} 
					while(V < 0)
							V = V + D;
					while(V >= D)
							V = V - D;
					V = V * 3.81972; 
					if(!i)
							sr = V; // sunrise
					else
							ss = V; // sunset
			} 
	   return {sr:sr,ss:ss,valid:invalid};
	}	
}


	

date/DATE.js
/* jdates.js               
 * Hebrew / Gregorian date conversion  
 *
 * December 13, 2004 - Tevet 1st, 5765 
 *  v4.67               	           
 * (c) Gabriel Zerbib,  gabriel@bumpt.net    http://www.bumpt.net     
 * It is strictly forbidden to use or  reproduce all or parts of this      
 * program without author's explicit   permission.                         
 * Commercial use of this program is   subject to purchase. Please contact 
 * the author.                         
 * abstract class JDate
 * JDate is the base class for GDate and HDate.
 * Do not instanciate JDate.
 * usage :
 * GD = new GDate(17,7,1984)
 * HD = new HDate(GD) */
function JDate(){
	/* runtime class info. Subclasses must implement (ptr to subclass construction function) */
	this.Class = null;
	/* * public methods */
	this.getDay = getDay;			/* () */
	this.getMonth = getMonth;		/* () */
	this.getYear = getYear;				/* () */
	this.getDayOfWeek = getDayOfWeek;	/* () */
	this.getWeekDayName = getWeekDayName 
	this.setDay = setDay;				/* (int) */
	this.setMonth = setMonth;			/* (int) */
	this.setYear = setYear;				/* (int) */
	this.eq = eq;				/* (JDate) */
	this.neq = neq;				/* (JDate) */
	this.lt = lt;				/* (JDate) */
	this.lte = lte;				/* (JDate) */
	this.gt = gt;				/* (JDate) */
	this.gte = gte;				/* (JDate) */
	this.add = add;				/* (int) */
	this.sub = sub;				/* (int) */
	this.minus = minus;			/* (JDate) */
	this.nextMonth = nextMonth;	/* () */
	this.prevMonth = prevMonth;	/* () */
	this.nextDay = nextDay;		/* () */
	this.setDate = setDate;		/* (), (int), (JDate), (int, int, int) */
	/* public pure virtual methods Subclasses must implement */
	this.today = null;
	this.getMonthLength = null;
	this.getYearLength = null;
	this.nbMonth = null;
	/*protected pure virtual methods * Subclasses must implement.	 * Do not call these methods outside the class hierarchy.*/
	this.calculate = null;
	this.setDate_int_int_int = null;
	/* protected attributes Do not access (read/write) these attributes outside the class hierarchy.*/
	this.m_hdn = 0;
	this.m_day = 0;
	this.m_month = 0;
	this.m_year = 0;
	/* protected methods Do not call these methods outside the class hierarchy. */
	this.setHdn = setHdn;
	this.setDate_int = setDate_int;
	this.setDate_JDate = setDate_JDate;
	/* Implementation */
	/* private methods */
	function setHdn(hdn){
		hdn = parseInt(hdn);
		if(hdn < 1) hdn = 1;
		this.m_hdn = hdn;
		this.calculate();
		return this;
	}
	function setDate_int(hdn){		this.setHdn(hdn);		return this;	}
	function setDate_JDate(jdate){	return this.setDate_int(jdate.m_hdn);}
	/* public attribute accessors */
	function getDay() 				{ return this.m_day; }
	function getMonth()				{ return this.m_month; }
	function getYear()				{ return this.m_year; }
	function setDay(day)			{ return this.setDate_int_int_int(day, this.m_month, this.m_year); }
	function setMonth(month)	{ return this.setDate_int_int_int(this.m_day, month, this.m_year); }
	function setYear(year)		{ return this.setDate_int_int_int(this.m_day, this.m_month, year); }
	function getDayOfWeek()		{ return this.m_hdn % 7; }
	function getWeekDayName(){	return ["Sunday","Monday","Tuesday","Wendsay","Thursday","Friday","Sat.day"][this.getDayOfWeek()]}
	function setDate(){
		var args = arguments;
		if(args.length == 1)
			if(typeof(args[0].length) != 'undefined')
				args = args[0];

		switch(args.length){
			case 1:
				if(args[0].constructor == JDate)	//setDate(JDate)
					this.setDate_JDate(args[0]);
				else
					this.setDate_int(args[0]);			//setDate(int)
				break;
			case 3:
				this.setDate_int_int_int(args[0], args[1], args[2]);	//setDate(int, int, int)
				break;
			default:	//setDate()
				this.today();
		}
	}
	/* public comparison functions */
	function eq(jdate)	{ return this.m_hdn == jdate.m_hdn; }
	function neq(jdate) { return ! this.eq(jdate); }
	function lt(jdate)	{ return this.m_hdn < jdate.m_hdn; }
	function lte(jdate) { return ( this.eq(jdate) || this.lt(jdate) ); }
	function gt(jdate)	{ return this.m_hdn > jdate.m_hdn; }
	function gte(jdate) { return ( this.eq(jdate) || this.gt(jdate) ); }
	/* public operations */
	function add(days) 		{ return this.setHdn(this.m_hdn + parseInt(days)); }
	function sub(days) 		{ return this.add( - parseInt(days) ); }
	function minus(jdate)	{ return (this.m_hdn - jdate.m_hdn); }
	function nextMonth()	{
		var d = this.getDay();
		this.add(this.getMonthLength() - d + 1);
		if(this.getMonthLength() >= d)
			this.setDay(d);
		else
			this.setDay(this.getMonthLength());
		return this;
	}
	function prevMonth()	{
		var d = this.getDay();
		this.sub(d);
		if(this.getMonthLength() >= d)
			this.setDay(d);
		else
			this.setDay(this.getMonthLength());
		return this;
	}
	function nextDay()	/* Add 1 day : method optimized within same month. */	{
		if( this.getDay() < this.getMonthLength() )	{
			this.m_day ++;
			this.m_hdn ++;
		}else
			this.add(1);
		return this;
	}
}
/******** class GDate *********/
GDate.prototype = new JDate; /* extends JDate */
GDate.isLeapYear = function (year){
	return (year % 400 == 0) || ( (year % 4 == 0) && (year % 100 != 0) );
};
function GDate(){
	/* runtime class info */
	this.Class = GDate;
	/* public methods */
	this.today = today;										/* () */
	this.getMonthLength = getMonthLength;	/* () */
	this.getYearLength = getYearLength;		/* () */
	this.isLeap = isLeap;									/* () */
	this.nbMonths = nbMonths							/* () */
	
	
	this.getMonthName = getMonthName
	
	/* public overrides */
	this.setDate_int_int_int = setDate_int_int_int;
	/* protected overrides Do not call outside the class hierarchy. */
	this.calculate = calculate;	/* () */
	/* constructor */
	this.setDate(arguments);
	/****** Implementation *****/
	/* private class methods. */
	
	function getMonthName(){
		return ["","January","February","March","April","May","June","July","August","September","October","November","December"]
		[this.getMonth()] ;
	}
	
	function st_monthLength(month, year){
		switch(month){
		case 2: return (GDate.isLeapYear(year) ? 29 : 28);
		case 4: case 6: case 9: case 11: return 30;
		case 1: case 3: case 5: case 7: case 8: case 10: case 12: return 31;
		default: return null;
		}
	}
	/*private methods */
	function calculate(){
		var a = this.m_hdn + 380041;
		var b = Math.floor((4 * a + 3 ) / 146097);
		var c = a - Math.floor((146097 * b) / 4);
		var d = Math.floor((4 * c + 3) / 1461);
		var e = c - Math.floor((1461 * d / 4));
		var m = Math.floor((5 * e + 2) / 153);

		this.m_day = e - Math.floor((153 * m + 2) / 5) + 1;
		this.m_month = m + 3 - 12 * Math.floor(m / 10);
		this.m_year = 100 * b + d - 4800 + Math.floor(m / 10);
	}
	/* public methods */
	function isLeap(){	return GDate.isLeapYear(this.m_year);	}
	function getMonthLength(){	return st_monthLength(this.m_month, this.m_year);	}
	function getYearLength()	{		return this.isLeap() ? 366 : 365;	}
	function today()	{		var d = new Date();	this.setDate_int_int_int(d.getDate(), d.getMonth() + 1, d.getFullYear());}
	function nbMonths()	{		return 12;	}
	/** protected overrides */
	function setDate_int_int_int(day, month, year){
		day = parseInt(day); month = parseInt(month); year = parseInt(year);
		if(month > 12) month = 12;
		var ml = st_monthLength(month, year);
		if(day > ml )
			day = ml;

		var a = Math.floor((14 - month) / 12);
		var y = year + 4800 - a;
		var m = month + 12 * a - 3;

		this.m_day = day;
		this.m_month = month;
		this.m_year = year;

		this.setHdn(day + Math.floor((153 * m + 2) / 5) + 365 * y + Math.floor(y / 4) - Math.floor(y / 100) + Math.floor(y / 400) - 380042);
		return this;
	}
}
/********* class HDate **      ********/
HDate.prototype = new JDate; /* extends JDate */
/* public static constants */
HDate.NISSAN = 1;
HDate.IYAR = 2;
HDate.SIVAN = 3;
HDate.TAMUZ = 4;
HDate.AV = 5;
HDate.ELUL = 6;
HDate.TISHRI = 7;
HDate.CHESHVAN = 8;
HDate.KISLEV = 9;
HDate.TEVET = 10;
HDate.SHVAT = 11;
HDate.ADAR = 12;
HDate.ADAR1 = 12;
HDate.ADAR2 = 13;

HDate.RISHON = 0;
HDate.CHAMISHI = 4;
HDate.SHISHI = 5;
HDate.SHABAT = 6;
HDate.SHABBAT = 6;


HDate.CHASERA = 1;
HDate.SDURA = 2;
HDate.SHLEMA = 3;
/* public static methods */
HDate.isEmbolismic = function(year){return (((12 * parseInt(year) + 17) % 19) >= 12);};
function HDate(){
	/* runtime class info */
	this.Class = HDate;
	/* public methods*/
	this.today = today;										/* () */
	this.getMonthLength = getMonthLength;	/* () */
	this.getYearLength = getYearLength;		/* () */
	this.isLeap = isLeap;									/* () */
	this.getYearType = getYearType;		/* () */
	this.nbMonths = nbMonths;		/* () */
	
	this.getMonthName = getMonthName 
	
	/* protected overrides */
	this.calculate = calculate;
	/* protected methods */
	this.setDate_int_int_int = setDate_int_int_int;
	/* private attributes */
	this.m_type = 0;
	/*	 * constructor	 */
	this.setDate(arguments);
	/***** * Implementation ******/
	/*	 * private class methods.	 */
	function st_embolismicYear(year)	{	return (((12 * parseInt(year) + 17) % 19) >= 12);}
	function st_roshHashanaDay(hyear){
		hyear = parseInt(hyear);
		var /*const*/ uday	= (24 * 1080);
		var /*const*/ Tsyn = 29 * uday + 12 * 1080 + 793;
		var /*const*/ Ttohu = 5 * 1080 + 204;
		var /*const*/ Tgatarad = 9 * 1080 + 204;
		var /*const*/ Tzaken = 18 * 1080;
		var /*const*/ Tbetutkafot = 15 * 1080 + 589;

		var Nmonths = Math.floor((235 * hyear - 234) / 19);
		var Tmolad = Nmonths * Tsyn + Ttohu;

		var days = Math.floor(Tmolad / uday);
		var parts = Tmolad - days * uday;
		var weekday = (1 + days) % 7;

		var adu = ( (weekday == 0) || (weekday == 3) || (weekday == 5) );
		var gatarad = ( !st_embolismicYear(hyear) && ((weekday == 2) && (parts >= Tgatarad)) );
		var betutkafot = ( (hyear == 1 ? false : (st_embolismicYear(hyear - 1))) && ((weekday == 1) && (parts >= Tbetutkafot)));

		var zaken = false;
		if(!adu && !gatarad && !betutkafot)	{
			zaken = (parts >= Tzaken);
			if(zaken) adu = ( (weekday == 2) || (weekday == 4) || (weekday == 6) );
		}
		return 1 + days + (adu?1:0) + (gatarad?2:0) + (betutkafot?1:0) + (zaken?1:0);
	}
	/* * private methods */
	function getMonthName(){
//		return ["","Nisan","Iyar","Sivan","Tamuz","Av","Elul","Tishri","Cheshvan","Kislev","Tevet","Shvat","Adar","Adar a","Adar b"]
		return ["","ניסן","אייר","סיוון","תמוז","אב","אלול","תשרי","חשוון","כסלו","טבת","שבט","אדר","אדר א","אדר ב"]
		[this.getMonth()] ;
	}
	function getHebNumber(){
//		return ["","Nisan","Iyar","Sivan","Tamuz","Av","Elul","Tishri","Cheshvan","Kislev","Tevet","Shvat","Adar","Adar a","Adar b"]
		return ["","א","ב","ג","ד","ה","ו","ז","ח","ט","י","יא","יב","יג","יד","טו","טז","יז","יח","יט","כ","כא","כב","כג","כד","כה","כו","כז","כח","כט","ל"];
	}
	function calculate(){
		var jdn;
		var days;
		var ml;

		var /*const*/ uday	= (24 * 1080);
		var /*const*/ Tsyn = 29 * uday + 12 * 1080 + 793;

		var hyear = 1 + Math.floor((234 + 19 * (this.m_hdn + 1) * uday / Tsyn) / 235);
		var rhnext = 0;
		var rh = st_roshHashanaDay(hyear);
		while(rh > this.m_hdn){
			rhnext = rh;
			hyear--;
			rh = st_roshHashanaDay(hyear);
		}

		if(rhnext == 0) rhnext = st_roshHashanaDay(hyear + 1);
		this.m_type = ((rhnext - rh) % 10) - 2;
		this.m_day = 1;
		this.m_month = HDate.TISHRI;
		this.m_year = hyear;

		days = this.m_hdn - rh;
		ml = this.getMonthLength();
		while(days >= ml)	{
			this.m_month = ( (this.m_month + 1 > this.nbMonths()) ? 1 : this.m_month + 1);
			days -= ml;
			ml = this.getMonthLength();
		}
		this.m_day += days;
	}

	/** public methods */
	function getYearType() { return this.m_type; }
	function isLeap() { return st_embolismicYear(this.m_year); }
	function nbMonths() { return (this.isLeap() ? 13 : 12); }

	function getMonthLength()	{
		var m = this.m_month;

		if(m == HDate.TISHRI) return 30;
		if(m == HDate.CHESHVAN) return (this.m_type == HDate.SHLEMA ? 30 : 29);
		if(m == HDate.KISLEV) return (this.m_type == HDate.CHASERA ? 29 : 30);
		if(m == HDate.TEVET) return 29;
		if(m == HDate.SHVAT) return 30;
		if(m == HDate.ADAR1) return (this.isLeap() ? 30 : 29);
		if(m == HDate.ADAR2) return (this.isLeap() ? 29 : null);
		if(m == HDate.NISSAN) return 30;
		if(m == HDate.IYAR) return 29;
		if(m == HDate.SIVAN) return 30;
		if(m == HDate.TAMUZ) return 29;
		if(m == HDate.AV) return 30;
		if(m == HDate.ELUL) return 29;

		return null;
	}

	function getYearLength(){
		switch(this.m_type)	{
			case HDate.SHLEMA: return (this.isLeap() ? 385 : 355);
			case HDate.SDURA: return (this.isLeap() ? 384 : 354);
			case HDate.CHASERA: return (this.isLeap() ? 383 : 353);
		}
	}

	function today() { this.setDate_JDate(new GDate()); }

	/* protected overrides */
	function setDate_int_int_int(day, month, year){
		day = parseInt(day); month = parseInt(month); year = parseInt(year);
	
		var nextRH;
		this.m_hdn = st_roshHashanaDay(year);
		nextRH = st_roshHashanaDay(year + 1);
		this.m_type = ((nextRH - this.m_hdn) % 10) - 2;
		this.m_day = 1;
		this.m_month = HDate.TISHRI;
		this.m_year = year;
		if(month > this.nbMonths()) month = this.nbMonths();
		while(this.m_month != month){
			this.m_hdn += this.getMonthLength();
			this.m_month = ( (this.m_month + 1 > (this.isLeap() ? 13 : 12)) ? 1 : this.m_month + 1);
		}
		if(day > this.getMonthLength()) day = this.getMonthLength();
		this.m_day = day;
		this.m_hdn += day - 1;
		if(this.m_hdn < 1) this.setHdn(this.m_hdn);
		return this;
	}
	
}
